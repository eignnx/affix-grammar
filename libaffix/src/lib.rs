#[macro_use]
extern crate serde_derive;

pub mod fault;
pub mod gen;
pub mod literate;
pub mod parser;
pub mod semantics;

#[cfg(test)]
mod tests {
    use super::*;
    use std::convert::TryFrom;
    use std::fs;
    use std::path::PathBuf;

    static RNG_SEED: u64 = 123456;
    static MAX_VALUES: usize = 100;

    #[test]
    fn run_snapshots() {
        let inputs = PathBuf::from(std::env!("CARGO_MANIFEST_DIR"))
            .join("src")
            .join("snapshot_inputs");

        for dir_entry in fs::read_dir(inputs).unwrap() {
            let path = dir_entry.unwrap().path();
            let src_text = fs::read_to_string(&path).unwrap();
            let parse_res = parser::syntax::ParsedGrammar::try_from(&src_text[..]);

            let parse_snapshot_name = format!(
                "PARSE_SNAPSHOT__{}",
                path.file_stem().unwrap().to_string_lossy()
            );
            insta::assert_json_snapshot!(parse_snapshot_name, parse_res);

            if let Some(grammar) = parse_res.ok() {
                let generator = gen::Generator::new_seeded(grammar, RNG_SEED);
                let generated = generate_n(generator, MAX_VALUES);
                let gen_snapshot_name = format!(
                    "GEN_SNAPSHOT__{}",
                    path.file_stem().unwrap().to_string_lossy()
                );
                insta::assert_json_snapshot!(gen_snapshot_name, generated);
            }
        }
    }

    fn generate_n<R: rand::Rng>(
        mut gen: gen::Generator<R>,
        n: usize,
    ) -> fault::DynamicRes<Vec<String>> {
        let mut sentences = vec![];
        for _ in 0..n {
            match gen.generate() {
                Ok(sentence) => sentences.push(sentence),
                Err(fault::DynamicErr::MaxTrialsExceeded { .. }) => return Ok(sentences),
                Err(e) => return Err(e),
            }
        }
        Ok(sentences)
    }
}
