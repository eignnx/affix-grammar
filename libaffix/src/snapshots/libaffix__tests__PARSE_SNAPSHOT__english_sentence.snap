---
source: libaffix/src/lib.rs
expression: parse_res
---
{
  "Ok": {
    "data_decls": [
      {
        "name": "Subject",
        "variants": {
          "dog": [],
          "cow": [],
          "marmot": [],
          "hedgehog": []
        }
      }
    ],
    "rule_decls": [
      {
        "signature": {
          "name": "start",
          "parameter_types": []
        },
        "cases": [
          {
            "guard": {
              "requirements": []
            },
            "alternatives": [
              [
                {
                  "RuleRef": {
                    "rule": "english_sentence",
                    "args": []
                  }
                }
              ]
            ]
          }
        ]
      },
      {
        "signature": {
          "name": "english_sentence",
          "parameter_types": []
        },
        "cases": [
          {
            "guard": {
              "requirements": []
            },
            "alternatives": [
              [
                {
                  "StrLit": "the"
                },
                {
                  "RuleRef": {
                    "rule": "subject",
                    "args": [
                      {
                        "Variable": [
                          "S",
                          ""
                        ]
                      }
                    ]
                  }
                },
                {
                  "StrLit": "eats"
                },
                {
                  "RuleRef": {
                    "rule": "object",
                    "args": []
                  }
                },
                "Plus",
                {
                  "StrLit": "."
                }
              ],
              [
                {
                  "StrLit": "the"
                },
                {
                  "RuleRef": {
                    "rule": "subject",
                    "args": [
                      {
                        "Variable": [
                          "S",
                          ""
                        ]
                      }
                    ]
                  }
                },
                {
                  "StrLit": "hits"
                },
                {
                  "RuleRef": {
                    "rule": "object",
                    "args": []
                  }
                },
                {
                  "StrLit": "with a bat."
                }
              ],
              [
                {
                  "StrLit": "the"
                },
                {
                  "RuleRef": {
                    "rule": "subject",
                    "args": [
                      {
                        "Variable": [
                          "S",
                          ""
                        ]
                      }
                    ]
                  }
                },
                {
                  "StrLit": "stumbles upon"
                },
                {
                  "RuleRef": {
                    "rule": "object",
                    "args": []
                  }
                },
                {
                  "RuleRef": {
                    "rule": "when",
                    "args": []
                  }
                },
                "Plus",
                {
                  "StrLit": "."
                }
              ],
              [
                {
                  "StrLit": "look at that"
                },
                {
                  "RuleRef": {
                    "rule": "subject",
                    "args": [
                      {
                        "Variable": [
                          "S",
                          "1"
                        ]
                      }
                    ]
                  }
                },
                {
                  "StrLit": "go. go"
                },
                {
                  "RuleRef": {
                    "rule": "subject",
                    "args": [
                      {
                        "Variable": [
                          "S",
                          "1"
                        ]
                      }
                    ]
                  }
                },
                "Plus",
                {
                  "StrLit": ", go."
                }
              ],
              [
                {
                  "StrLit": "the"
                },
                {
                  "RuleRef": {
                    "rule": "subject",
                    "args": [
                      {
                        "Variable": [
                          "S",
                          ""
                        ]
                      }
                    ]
                  }
                },
                {
                  "StrLit": "began setting"
                },
                {
                  "RuleRef": {
                    "rule": "object",
                    "args": []
                  }
                },
                {
                  "StrLit": "on fire."
                }
              ],
              [
                {
                  "RuleRef": {
                    "rule": "english_sentence",
                    "args": []
                  }
                },
                {
                  "RuleRef": {
                    "rule": "connective",
                    "args": []
                  }
                },
                "Plus",
                {
                  "StrLit": ","
                },
                {
                  "RuleRef": {
                    "rule": "english_sentence",
                    "args": []
                  }
                }
              ]
            ]
          }
        ]
      },
      {
        "signature": {
          "name": "when",
          "parameter_types": []
        },
        "cases": [
          {
            "guard": {
              "requirements": []
            },
            "alternatives": [
              [
                {
                  "StrLit": "while on their morning walk"
                }
              ],
              [
                {
                  "StrLit": "during their annual checkup"
                }
              ],
              [
                {
                  "StrLit": "upon stepping outdoors"
                }
              ],
              [
                {
                  "StrLit": "while mother is out"
                }
              ],
              [
                {
                  "StrLit": "under the cover of darkness"
                }
              ],
              [
                {
                  "StrLit": "at the corner store"
                }
              ]
            ]
          }
        ]
      },
      {
        "signature": {
          "name": "subject",
          "parameter_types": [
            "Subject"
          ]
        },
        "cases": [
          {
            "guard": {
              "requirements": [
                {
                  "Variant": "dog"
                }
              ]
            },
            "alternatives": [
              [
                {
                  "StrLit": "dog"
                }
              ]
            ]
          },
          {
            "guard": {
              "requirements": [
                {
                  "Variant": "cow"
                }
              ]
            },
            "alternatives": [
              [
                {
                  "StrLit": "cow"
                }
              ]
            ]
          },
          {
            "guard": {
              "requirements": [
                {
                  "Variant": "hedgehog"
                }
              ]
            },
            "alternatives": [
              [
                {
                  "StrLit": "hedgehog"
                }
              ]
            ]
          },
          {
            "guard": {
              "requirements": [
                {
                  "Variant": "marmot"
                }
              ]
            },
            "alternatives": [
              [
                {
                  "StrLit": "marmot"
                }
              ]
            ]
          }
        ]
      },
      {
        "signature": {
          "name": "object",
          "parameter_types": []
        },
        "cases": [
          {
            "guard": {
              "requirements": []
            },
            "alternatives": [
              [
                {
                  "StrLit": "the chair"
                }
              ],
              [
                {
                  "StrLit": "the couch"
                }
              ],
              [
                {
                  "StrLit": "the"
                },
                {
                  "RuleRef": {
                    "rule": "subject",
                    "args": [
                      {
                        "Variable": [
                          "S",
                          ""
                        ]
                      }
                    ]
                  }
                },
                "Plus",
                {
                  "StrLit": "'s lunch"
                }
              ],
              [
                {
                  "StrLit": "an extremely loud"
                },
                {
                  "RuleRef": {
                    "rule": "subject",
                    "args": [
                      {
                        "Variable": [
                          "S",
                          ""
                        ]
                      }
                    ]
                  }
                }
              ],
              [
                {
                  "StrLit": "an obscene amount of"
                },
                {
                  "RuleRef": {
                    "rule": "uncountable_thing",
                    "args": []
                  }
                }
              ],
              [
                {
                  "StrLit": "a steaming pile of"
                },
                {
                  "RuleRef": {
                    "rule": "uncountable_thing",
                    "args": []
                  }
                }
              ],
              [
                {
                  "StrLit": "a lifetime supply of"
                },
                {
                  "RuleRef": {
                    "rule": "uncountable_thing",
                    "args": []
                  }
                }
              ]
            ]
          }
        ]
      },
      {
        "signature": {
          "name": "uncountable_thing",
          "parameter_types": []
        },
        "cases": [
          {
            "guard": {
              "requirements": []
            },
            "alternatives": [
              [
                {
                  "StrLit": "jelly"
                }
              ],
              [
                {
                  "StrLit": "tuna"
                }
              ],
              [
                {
                  "StrLit": "bees"
                }
              ],
              [
                {
                  "StrLit": "fruit-flies"
                }
              ],
              [
                {
                  "StrLit": "medical bills"
                }
              ],
              [
                {
                  "StrLit": "caramel"
                }
              ],
              [
                {
                  "StrLit": "ketchup and mustard"
                }
              ]
            ]
          }
        ]
      },
      {
        "signature": {
          "name": "connective",
          "parameter_types": []
        },
        "cases": [
          {
            "guard": {
              "requirements": []
            },
            "alternatives": [
              [
                {
                  "StrLit": "unfortunately"
                }
              ],
              [
                {
                  "StrLit": "but later that day"
                }
              ],
              [
                {
                  "StrLit": "with the greatest of ease"
                }
              ],
              [
                {
                  "StrLit": "with a twitch of it's nose"
                }
              ],
              [
                {
                  "StrLit": "in the blink of an eye"
                }
              ],
              [
                {
                  "StrLit": "having completed the task at hand"
                }
              ],
              [
                {
                  "StrLit": "having exhausted their options"
                }
              ]
            ]
          }
        ]
      }
    ]
  }
}
